packages/factory/src/helpers.ts

```diff
@@ -3,32 +3,100 @@
 // Copyright 2019-Present Datadog, Inc.
 
 import type {
+    BuildReport,
     BundlerFullName,
     BundlerName,
     FactoryMeta,
+    GetLogger,
     GlobalContext,
+    LogLevel,
     Options,
+    OptionsWithDefaults,
     ToInjectItem,
 } from '@dd/core/types';
+import c from 'chalk';
+
+const logPriority: Record<LogLevel, number> = {
+    debug: 0,
+    info: 1,
+    warn: 2,
+    error: 3,
+    none: 4,
+};
+
+// Exported for testing.
+export const getLoggerFactory =
+    (build: BuildReport, logLevel: LogLevel = 'warn'): GetLogger =>
+    (name) => {
+        const log = (text: any, type: LogLevel = 'debug') => {
+            // By default (debug) we print dimmed.
+            let color = c.dim;
+            let logFn = console.log;
+
+            if (type === 'error') {
+                color = c.red;
+                logFn = console.error;
+            } else if (type === 'warn') {
+                color = c.yellow;
+                logFn = console.warn;
+            } else if (type === 'info') {
+                color = c.cyan;
+                logFn = console.log;
+            }
+
+            const prefix = `[${type}|${name}]`;
+
+            // Keep a trace of the log in the build report.
+            const content = typeof text === 'string' ? text : JSON.stringify(text, null, 2);
+            build.logs.push({ pluginName: name, type, message: content, time: Date.now() });
+            if (type === 'error') {
+                build.errors.push(content);
+            }
+            if (type === 'warn') {
+                build.warnings.push(content);
+            }
+
+            // Only log if the log level is high enough.
+            if (logPriority[type] >= logPriority[logLevel]) {
+                logFn(`${color(prefix)} ${content}`);
+            }
+        };
+
+        const logFn = (text: any, type: LogLevel = 'debug') => {
+            log(text, type);
+        };
```

Comment:
I personally find it a bit weird to have both a top level fn and `logger.<level>()` fns